src/StableMemory.mo

```diff
@@ -0,0 +1,373 @@
+/// (matching and reusing a recent Rust library-version of the same feature).
+///
+/// The API of the `class` provides the same low-level API as before, but where each instance
+/// can be used independently of the others.
+///
+/// Reclaiming space from one to use in another is
+/// unlocked by this new generalized API, but is not yet implemented.
+/// By creating a more sophisticated backing allocator, future implementations
+/// (more complex than the present one lifted from the Rust library)
+/// could support freeing memories transparently via the existing GC infrastructure,
+/// avoid growing the underlying canister stable memory when a new instance is required,
+/// or when an existing instance must grow.
+///
+/// ## Per-instance API
+///
+/// Memory is allocated, using `grow(pages)`, sequentially and on demand, in units of 64KiB pages, starting with 0 allocated pages.
+/// New pages are zero initialized.
+/// Growth is capped by a soft limit on page count controlled by compile-time flag
+/// `--max-stable-pages <n>` (the default is 65536, or 4GiB).
+///
+/// Each `load` operation loads from byte address `offset` in little-endian
+/// format using the natural bit-width of the type in question.
+/// The operation traps if attempting to read beyond the current stable memory size.
+///
+/// Each `store` operation stores to byte address `offset` in little-endian format using the natural bit-width of the type in question.
+/// The operation traps if attempting to write beyond the current stable memory size.
+///
+/// Text values can be handled by using `Text.decodeUtf8` and `Text.encodeUtf8`, in conjunction with `loadBlob` and `storeBlob`.
+///
+/// The current page allocation and page contents is preserved across upgrades.
+///
+/// NB: The IC's actual stable memory size (`ic0.stable_size`) may exceed the
+/// page size reported by Motoko function `size()`.
+/// This (and the cap on growth) are to accommodate Motoko's stable variables.
+/// Applications that plan to use Motoko stable variables sparingly or not at all can
+/// increase `--max-stable-pages` as desired, approaching the IC maximum (initially 8GiB, then 32Gib, currently 48Gib).
+/// All applications should reserve at least one page for stable variable data, even when no stable variables are used.
+///
+/// Usage:
+/// ```motoko no-repl
+/// import StableMemory "mo:base/ExperimentalStableMemory";
+/// ```
+
+import Prim "mo:⛔";
+
+module {
+
+  public class StableMemory () {
+
+  var id = Prim.multStableMemoryNew();
+
+  /// Current size of the stable memory, in pages.
+  /// Each page is 64KiB (65536 bytes).
+  /// Initially `0`.
+  /// Preserved across upgrades, together with contents of allocated
+  /// stable memory.
+  ///
+  /// Example:
+  /// ```motoko no-repl
+  /// let beforeSize = StableMemory.size();
+  /// ignore StableMemory.grow(10);
+  /// let afterSize = StableMemory.size();
+  /// afterSize - beforeSize // => 10
+  /// ```
```

Comment:
```Suggestion:
/// let memory = StableMemory();
/// let beforeSize = memory.size();
/// ignore memory.grow(10);
/// let afterSize = memory.size();
/// afterSize - beforeSize // => 10

Reply:
Just a few thoughts (not sure whether relevant or useful):
* Maybe we should abstract the page sizes and memory address space size away, to ease a potential future upgrade to 64-bit.
* The attribute "stable" confused me at the beginning, because it might suggest that other state of Motoko is not stable. Maybe, it could be named "raw memory" or "raw stable memory". (Ideally, if we would aim for true orthogonal persistence with scalable upgrades, where there would be no need for explicit distinction between stable and flexible/non-stable state.)
* As proposed, stringent bounds checking seems like a very important aspect to me.
* As @crusso suggests, some abstraction of a synthetic pointer to stable memory would be useful, such that one can reference across different stable memory instances. I guess a class-based approach could work if it encapsulates the inner raw location and returns the pointer as class instances for convenience.

```
// not necessarily a stable type
class RawPointer {
  memory: RawMemory;
  offset: Nat64;
}

// designed as a stable type
class RawMemory {
  // internal representation: e.g. addresses of the pages assigned by the allocator.
  public func size(): Nat;
  public func pointTo(offset: Nat64): RawPointer;
}

func allocate(size: Nat): RawMemory;

func loadNat32(location: RawPointer): Nat32;
func storeNat32(location: RawPointer, value: Nat32);

func loadPointer(location: RawPointer): RawPointer; 
func storePointer(location: RawPointer, value: RawPointer);
```

Maybe, a source of inspiration could be the Java native access API.

Reply:
I agree, good point. Probably, it would be simpler with immediate offsets (like JNA).